subprojects/execution/src/main/java/org/gradle/internal/execution/workspace/impl/DefaultImmutableWorkspaceProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.workspace.impl;
+
+import org.gradle.api.internal.cache.StringInterner;
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.CleanupAction;
+import org.gradle.cache.FileLockManager;
+import org.gradle.cache.PersistentCache;
+import org.gradle.cache.internal.CompositeCleanupAction;
+import org.gradle.cache.internal.InMemoryCacheDecoratorFactory;
+import org.gradle.cache.internal.LeastRecentlyUsedCacheCleanup;
+import org.gradle.cache.internal.SingleDepthFilesFinder;
+import org.gradle.internal.execution.history.ExecutionHistoryStore;
+import org.gradle.internal.execution.history.impl.DefaultExecutionHistoryStore;
+import org.gradle.internal.execution.workspace.ImmutableWorkspaceProvider;
+import org.gradle.internal.file.FileAccessTimeJournal;
+import org.gradle.internal.file.impl.SingleDepthFileAccessTracker;
+
+import java.io.Closeable;
+import java.io.File;
+
+import static org.gradle.cache.internal.LeastRecentlyUsedCacheCleanup.DEFAULT_MAX_AGE_IN_DAYS_FOR_RECREATABLE_CACHE_ENTRIES;
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
+public class DefaultImmutableWorkspaceProvider implements ImmutableWorkspaceProvider, Closeable {
+    private static final int FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP = 1;
+
+    private final SingleDepthFileAccessTracker fileAccessTracker;
+    private final File baseDirectory;
+    private final ExecutionHistoryStore executionHistoryStore;
+    private final PersistentCache cache;
+
+    public DefaultImmutableWorkspaceProvider(
+        String name,
+        File baseDirectory,
+        CacheRepository cacheRepository,
+        FileAccessTimeJournal fileAccessTimeJournal,
+        InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory,
+        StringInterner stringInterner
+    ) {
+        this.baseDirectory = baseDirectory;
+        this.cache = cacheRepository
+            .cache(baseDirectory)
+            .withCleanup(createCleanupAction(baseDirectory, fileAccessTimeJournal))
+            .withDisplayName(name)
+            .withLockOptions(mode(FileLockManager.LockMode.OnDemand)) // Lock on demand
+            .open();
```

Comment:
@wolfs can you please take a look if the parameters here make sense to you?

Reply:
Is there already a test for the cleanup? If there isn't, I think you should add one.